src/metabase/prometheus.clj

```diff
@@ -0,0 +1,127 @@
+            [metabase.models.setting :as setting :refer [defsetting]]
+            [metabase.server :as server]
+            [metabase.util.i18n :refer [trs]]
+            [potemkin :as p]
+            [potemkin.types :as p.types]
+            [ring.adapter.jetty :as ring-jetty])
+  (:import [io.prometheus.client.hotspot
+            GarbageCollectorExports
+            MemoryPoolsExports
+            StandardExports
+            ThreadExports]
+           io.prometheus.client.jetty.JettyStatisticsCollector
+           org.eclipse.jetty.server.Server))
+
+(defsetting prometheus-server-port
+  "Port to serve prometheus status from. If set"
+  :type       :integer
+  :visibility :internal
+  ;; settable only through environmental variable
+  :setter     :none
+  :getter     (fn reading-prometheus-port-setting []
+                (let [parse (fn [raw-value]
+                              (if-let [parsed (parse-long raw-value)]
+                                parsed
+                                (log/warn (trs "MB_PROMETHEUS_SERVER_PORT value of ''{0}'' is not parseable as an integer."
+                                               raw-value))))]
+                  (setting/get-raw-value :prometheus-server-port integer? parse))))
+
+(defonce ^:private ^{:doc "Prometheus System for prometheus metrics"} system nil)
+
+(defn- jvm-collectors
+  "JVM collectors. Essentially duplicating [[iapetos.collector.jvm]] namespace so we can set our own namespaces rather
+  than \"iapetos_internal\""
+  []
+  [(collector/named {:namespace "metabase_application"
+                     :name      "jvm_gc"}
+                    (GarbageCollectorExports.))
+   (collector/named {:namespace "metabase_application"
+                     :name      "jvm_standard"}
+                    (StandardExports.))
+   (collector/named {:namespace "metabase_application"
+                     :name      "jvm_memory_pools"}
+                    (MemoryPoolsExports.))
+   (collector/named {:namespace "metabase_application"
+                     :name      "jvm_threads"}
+                    (ThreadExports.))])
+
+(defn- jetty-collectors
+  []
+  [(JettyStatisticsCollector. (.getHandler (server/instance)))])
+
+(defn- setup-metrics!
+  "Instrument the application. Conditionally done when some setting is set. If [[prometheus-server-port]] is not set it
+  will throw."
+  [registry-name]
+  (log/info (trs "Starting prometheus metrics collector"))
+  (let [registry (prometheus/collector-registry registry-name)]
+    (apply prometheus/register registry
+           (concat (jvm-collectors)
+                   (jetty-collectors)))))
+
+(defn- start-web-server!
+  "Start the prometheus web-server. If [[prometheus-server-port]] is not set it will throw."
+  [port registry]
+  (log/info (trs "Starting prometheus metrics web-server on port {0}" port))
+  (when-not port
+    (throw (ex-info (trs "Attempting to set up prometheus metrics web-server with no web-server port provided")
```

Comment:
```suggestion
    (throw (ex-info (trs "Failed to set up prometheus metrics web-server with no web-server port provided")
```

IMO "Attempting" gives the impression of starting to setup, not something went wrong.